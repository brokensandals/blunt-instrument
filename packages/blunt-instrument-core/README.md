# blunt-instrument-core

Contents:

- [ASTBundle](#astbundle)
- [Tracer](#tracer)
- [ArrayTrace](#arraytrace)
- [TrevCollection](#trevcollection)

## ASTBundle

Programs that have been instrumented with blunt-instrument record the value of each expression they evaluate.
Every value is associated with the node of the program's abstract syntax tree corresponding to the expression, so that you can look up all the values for any given expression.
blunt-instrument uses the [ASTs produced by babel][babel-ast].
This class wraps a set of those ASTs and adds some extra fields to them.

(The instrumenter can instrument multiple files so that they'll produce a combined trace; each file's AST will have its own astId.)

```javascript
const astb = new ASTBundle();
astb.add('one', ast1, sourceCode1);
astb.add('two', ast2, sourceCode2);
astb.getNode('one', 10);
astb.getNodeByKey(toNodeKey('one', 10));
astb.filterNodes((node) => node.codeSlice === 'x + 1');
```

See the JSDoc in [ASTBundle.js](src/ast/ASTBundle.js) for more info.

### Node IDs

Every node in the AST needs a unique identifier so that the tracer can indicate which node a result corresponds to.
This is stored in the `biId` field.

To assign sequential node IDs to all nodes in a tree that don't have them, use `addIdsToAST`.
blunt-instrument-babel-plugin calls this for you.

`biId` is only unique within a single AST.
You need to know the AST's id too to fully identify a node.
The functions `toNodeKey` and `fromNodeKey` provide a way to encode/decode the combination of AST id and node id into a single string for convenience.

`copyNodeIdsBetweenASTs` is available for less common situations.

## Tracer

Instrumented code must be given an instance of the `Tracer` class.
You can either create an instance via `new Tracer()` or rely on the `defaultTracer` instance exported by the package.
If you're using [blunt-instrument-eval][eval], it will create an instance for you; if you're using the [babel plugin][babel-plugin] directly, you can configure it as documented there.

The `Tracer` supports adding listeners which will receive two callbacks, described below.
These can be used for recording the data of a trace; see `ArrayTrace` below for a basic implementation.

### handleTrev

This will be called every time the instrumented code reports that a tracing event ("trev") occurred.
These events include the evaluation of an expression, that start of a function call, returning from a function call, etc.

```js
tracer.addListener({
  onTrev(trev) {
    console.log(trev);
  }
});
```

See [babel-plugin-blunt-instrument's README][babel-plugin] for an example of what trevs look like.
Unlike what's shown there, though, the `data` field on this trev object has not been cloned or encoded in any way, so it might be modified as soon as control returns to the instrumented code.
(The `ArrayTracer` class replaces the `data` field with an encoded copy.)

### handleRegisterAST

This will be called when the instrumented code is initializing, unless that functionality has been disabled in the babel plugin opts.
The arguments are the AST ID (important if you have instrumented multiple files that are all reporting to the same Tracer) and the root babel Node of the AST.

```js
tracer.addListener({
  onRegisterAST(astId, ast) {
    console.log(astId, ast);
  }
});
```

## ArrayTrace

An `ArrayTrace` instance will listen to all the trevs reported by a `Tracer` instance and save them to an array.
It also keeps track of any ASTs registered via the Tracer.
See [blunt-instrument-eval][eval] for a simplified way to instrument & run code & retrieve an ArrayTrace from it.

```js
import { ArrayTrace } from 'blunt-instrument-trace-utils';

const trace = new ArrayTrace();
tracer.addListener(trace);
/* ... invoke some instrumented code ... */

console.log(trace.trevs);
```

See the jsdoc in [ArrayTrace.js](src/trace/ArrayTrace.js) for more info.

See the [babel plugin's README][babel-plugin] for an example of what an array of trevs looks like.

### Data Encoding

One difficulty in producing a full trace of a program is that every value (every number and string, every state of every array and object) needs to be copied or serialized in some way.
`ArrayTrace` encodes the `data` field in every trev using [object-graph-as-json][object-graph-as-json], which attempts to preserve as much information about the objects being copied as it can.

By default, each instance of `ArrayTrace` creates a new instance of `Encoder`, but you can specify an encoder to the constructor instead if you wish:

```js
const trace = new ArrayTrace({ encoder: myEncoder });
```

## TrevCollection

A TrevCollection contains a list of trace events, along with the ASTs for the code that generated those events.
It provides convenience methods for denormalizing extra data onto the trevs and getting basic statistics about them.

```javascript
const tc = arrayTrace.toTC();

// Get all the trevs in the collection
trevCollection.trevs;

// Get a new TrevCollection filtered according to a function you provide -
// in this case, find all the times that any expression evaluated to the
// string "hello world!"
tc.filter((trev) => trev.type === 'expr' && trev.data === 'hello world!');

// Attach extra fields to all the trevs for convenience
const fancyTC = tc.withDenormalizedInfo();

// Filter out literals; i.e., for a trace of the code 'x = y + 3',
// omit all the entries reporting that 3 evaluated to 3.
import types from '@babel/types';
fancyTC.filter((trev) => !types.isLiteral(trev.denormalized.node));

// Get the number of trevs by node, node type, and type.
const { nodes, nodeTypes, types } = tc.getFacets();
```

See the JSDoc in [TrevCollection.js](src/trace/TrevCollection.js) for more info.

[babel-ast]: https://github.com/jamiebuilds/babel-handbook/blob/master/translations/en/plugin-handbook.md#toc-asts
[eval]: ../blunt-instrument-eval/README.md
[babel-plugin]: ../babel-plugin-blunt-instrument/README.md
[object-graph-as-json]: https://github.com/brokensandals/object-graph-as-json
