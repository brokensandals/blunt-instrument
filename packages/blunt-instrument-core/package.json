{
  "name": "blunt-instrument-core",
  "version": "0.1.0",
  "description": "Utilities for recording and managing code traces.",
  "main": "target/cjs/index.js",
  "scripts": {
    "build:cjs": "BABEL_MODULES=cjs babel src/ --out-dir target/cjs/ --ignore src/ast/*.test.js,src/trace/*.test.js --root-mode upward",
    "build:esm": "BABEL_MODULES=false babel src/ --out-dir target/esm/ --ignore src/ast/*.test.js,src/trace/*.test.js --root-mode upward",
    "build": "npm run build:cjs && npm run build:esm",
    "clean": "rimraf target"
  },
  "devDependencies": {
    "@babel/cli": "^7.7.5",
    "@babel/core": "^7.7.5"
  },
  "dependencies": {
    "lodash": "^4.17.15",
    "object-graph-as-json": "^2.0.0"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/brokensandals/blunt-instrument.git"
  },
  "author": "Jacob Williams <jacobaw@gmail.com> (https://brokensandals.net)",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/brokensandals/blunt-instrument/issues"
  },
  "homepage": "https://github.com/brokensandals/blunt-instrument#readme"
}
