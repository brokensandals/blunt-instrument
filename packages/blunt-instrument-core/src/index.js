import addNodeIdsToAST from './ast/addNodeIdsToAST';
import ASTBundle from './ast/ASTBundle';
import attachCodeSlicesToAST from './ast/attachCodeSlicesToAST';
import copyNodeIdsBetweenASTs from './ast/copyNodeIdsBetweenASTs';
import fromNodeKey from './ast/fromNodeKey';
import toNodeKey from './ast/toNodeKey';

import ArrayTrace from './trace/ArrayTrace';
import ConsoleTraceWriter from './trace/ConsoleTraceWriter';
import Tracer from './trace/Tracer';
import TrevCollection from './trace/TrevCollection';

const defaultTracer = new Tracer();

export {
  addNodeIdsToAST,
  ASTBundle,
  attachCodeSlicesToAST,
  copyNodeIdsBetweenASTs,
  fromNodeKey,
  toNodeKey,
  ArrayTrace,
  ConsoleTraceWriter,
  Tracer,
  TrevCollection,
  defaultTracer,
};
