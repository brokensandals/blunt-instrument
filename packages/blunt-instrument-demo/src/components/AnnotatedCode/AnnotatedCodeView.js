import React from 'react';
import './AnnotatedCode.css';
import { getCodeSlice, getNodeId } from 'blunt-instrument-ast-utils';

// TODO babel probably provides a better way to do this
function findNodes(object) {
  if (object == null || typeof object !== 'object') {
    return [];
  }

  if (Array.isArray(object)) {
    return object.flatMap(findNodes);
  }

  if (object.type) {
    return [object];
  }

  return Object.keys(object).flatMap(k => findNodes(object[k]));
}

function findChildNodes(node) {
  return Object.keys(node).flatMap(k => findNodes(node[k]));
}

function CodeForNode({
  code,
  highlightedNodeId,
  node,
  onHoveredNodeChange,
  onNodeSelectedToggle,
  selectedNodeIds,
}) {
  const { extra: { biNodeId: nodeId }, start, end } = node;
  let cur = start;
  const elements = [];

  // TODO does this sorting need to look at 'end' too?
  const children = findChildNodes(node).sort((a, b) => a.start - b.start);
  for (const child of children) {
    if (cur < child.start) {
      elements.push(code.slice(cur, child.start));
    }
    
    elements.push(
      <CodeForNode key={getNodeId(child)}
                     highlightedNodeId={highlightedNodeId}
                     node={child}
                     selectedNodeIds={selectedNodeIds}
                     code={code}
                     onHoveredNodeChange={onHoveredNodeChange}
                     onNodeSelectedToggle={onNodeSelectedToggle} />
    );

    cur = child.end;
  }

  if (cur < end) {
    elements.push(code.slice(cur, end));
  }

  const handleMouseOver = onHoveredNodeChange ? (event) => {
    onHoveredNodeChange(nodeId);
    event.stopPropagation();
  } : null;

  const handleClick = onNodeSelectedToggle ? (event) => {
    onNodeSelectedToggle(nodeId);
    event.stopPropagation();
  } : null;

  const className = [
    nodeId === highlightedNodeId ? 'highlighted' : '',
    selectedNodeIds.includes(nodeId) ? 'selected' : '',
  ].join(' ')

  return <span className={className}
               onClick={handleClick}
               onMouseOver={handleMouseOver}>
    {elements}
  </span>
}

function AnnotatedCodeView({
  astQuerier,
  highlightedNodeId,
  onHoveredNodeChange = null,
  onNodeSelectedToggle = null,
  selectedNodeIds,
}) {
  const clearHover = onHoveredNodeChange ? () => onHoveredNodeChange(null) : null;
  return (
    <pre className="AnnotatedCode" onMouseLeave={clearHover}>
      <code>
        <CodeForNode highlightedNodeId={highlightedNodeId}
                       node={astQuerier.ast}
                       selectedNodeIds={selectedNodeIds}
                       code={getCodeSlice(astQuerier.ast)}
                       onHoveredNodeChange={onHoveredNodeChange}
                       onNodeSelectedToggle={onNodeSelectedToggle} />
      </code>
    </pre>
  );
}

export default AnnotatedCodeView;