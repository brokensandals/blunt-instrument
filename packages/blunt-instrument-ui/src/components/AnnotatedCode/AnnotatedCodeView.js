import React from 'react';
import './AnnotatedCode.css';

// TODO babel probably provides a better way to do this
function findNodes(object) {
  if (object == null || typeof object !== 'object') {
    return [];
  }

  if (Array.isArray(object)) {
    return object.flatMap(findNodes);
  }

  if (object.type) {
    return [object];
  }

  return Object.keys(object).flatMap(k => findNodes(object[k]));
}

function findChildNodes(node) {
  return Object.keys(node).flatMap(k => findNodes(node[k]));
}

function CodeForNode({
  code,
  highlightedNodeKey,
  node,
  onHoveredNodeChange,
  onNodeSelectedToggle,
  selectedNodeKeys,
}) {
  const { biKey: nodeKey, start, end } = node;
  let cur = start;
  const elements = [];

  // For code using shorthand object property notation, i.e. `{x}` instead of `{x: x}`,
  // there are three nodes corresponding to the code `x`: the ObjectProperty, ObjectProperty.id,
  // and ObjectProperty.value. This special case is here to avoid displaying `x` multiple times,
  // and to make sure the one we display is the one that trevs will actually be attached to.
  if (node.type === 'ObjectProperty' && node.shorthand) {
    return <CodeForNode key={node.value.biKey}
                        highlightedNodeKey={highlightedNodeKey}
                        node={node.value}
                        selectedNodeKeys={selectedNodeKeys}
                        code={code}
                        onHoveredNodeChange={onHoveredNodeChange}
                        onNodeSelectedToggle={onNodeSelectedToggle} />;
  }

  // TODO does this sorting need to look at 'end' too?
  const children = findChildNodes(node).sort((a, b) => a.start - b.start);
  for (const child of children) {
    if (cur < child.start) {
      elements.push(code.slice(cur, child.start));
    }
    
    elements.push(
      <CodeForNode key={child.biKey}
                     highlightedNodeKey={highlightedNodeKey}
                     node={child}
                     selectedNodeKeys={selectedNodeKeys}
                     code={code}
                     onHoveredNodeChange={onHoveredNodeChange}
                     onNodeSelectedToggle={onNodeSelectedToggle} />
    );

    cur = child.end;
  }

  if (cur < end) {
    elements.push(code.slice(cur, end));
  }

  const handleMouseOver = nodeKey && onHoveredNodeChange ? (event) => {
    onHoveredNodeChange(nodeKey);
    event.stopPropagation();
  } : null;

  const handleClick = nodeKey && onNodeSelectedToggle ? (event) => {
    onNodeSelectedToggle(nodeKey);
    event.stopPropagation();
  } : null;

  const className = [
    nodeKey && nodeKey === highlightedNodeKey ? 'highlighted' : '',
    nodeKey && selectedNodeKeys.includes(nodeKey) ? 'selected' : '',
  ].join(' ')

  return <span className={className}
               onClick={handleClick}
               onMouseOver={handleMouseOver}>
    {elements}
  </span>
}

function AnnotatedCodeView({
  ast,
  highlightedNodeKey,
  onHoveredNodeChange = null,
  onNodeSelectedToggle = null,
  selectedNodeKeys,
}) {
  const clearHover = onHoveredNodeChange ? () => onHoveredNodeChange(null) : null;
  return (
    <pre className="AnnotatedCode" onMouseLeave={clearHover}>
      <code>
        <CodeForNode highlightedNodeKey={highlightedNodeKey}
                       node={ast}
                       selectedNodeKeys={selectedNodeKeys}
                       code={ast.codeSlice}
                       onHoveredNodeChange={onHoveredNodeChange}
                       onNodeSelectedToggle={onNodeSelectedToggle} />
      </code>
    </pre>
  );
}

export default AnnotatedCodeView;