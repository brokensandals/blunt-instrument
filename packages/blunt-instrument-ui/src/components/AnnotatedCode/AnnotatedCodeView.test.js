import React from 'react';
import renderer from 'react-test-renderer';
import { parseSync } from '@babel/core';
import { addNodeIdsToAST, ASTBundle } from 'blunt-instrument-core';
import AnnotatedCodeView from './AnnotatedCodeView';

describe('AnnotatedCodeView', () => {
  const code = `function fac(n) {
    return n === 1 ? 1 : n * fac(n - 1);
  }`;
  let ast;
  let astb;

  beforeEach(() => {
    ast = parseSync(code);
    addNodeIdsToAST(ast);
    astb = new ASTBundle();
    astb.add('test', ast, code);
    ast = astb.asts.test;
  })

  test('nothing highlighted or selected', () => {
    const result = renderer.create(<AnnotatedCodeView ast={ast} />).toJSON();
    expect(result).toMatchSnapshot();
  });

  test('with a node highlighted', () => {
    const result = renderer.create(
      <AnnotatedCodeView ast={ast} highlightedNodeKey="test:5" />).toJSON();
    expect(result).toMatchSnapshot();
  });

  test('with nodes selected', () => {
    const result = renderer.create(
      <AnnotatedCodeView ast={ast} selectedNodeKeys={['test:5', 'test:8']} />).toJSON();
    expect(result).toMatchSnapshot();
  });
});
