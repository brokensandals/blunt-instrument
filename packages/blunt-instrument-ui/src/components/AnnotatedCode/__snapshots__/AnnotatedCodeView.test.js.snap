// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnnotatedCodeView nothing highlighted or selected 1`] = `
<pre
  className="AnnotatedCode"
  onMouseLeave={null}
>
  <code>
    <span
      className=" "
      onClick={null}
      onMouseOver={null}
    >
      <span
        className=" "
        onClick={null}
        onMouseOver={null}
      >
        <span
          className=" "
          onClick={null}
          onMouseOver={null}
        >
          function 
          <span
            className=" "
            onClick={null}
            onMouseOver={null}
          >
            fac
          </span>
          (
          <span
            className=" "
            onClick={null}
            onMouseOver={null}
          >
            n
          </span>
          ) 
          <span
            className=" "
            onClick={null}
            onMouseOver={null}
          >
            {
    
            <span
              className=" "
              onClick={null}
              onMouseOver={null}
            >
              return 
              <span
                className=" "
                onClick={null}
                onMouseOver={null}
              >
                <span
                  className=" "
                  onClick={null}
                  onMouseOver={null}
                >
                  <span
                    className=" "
                    onClick={null}
                    onMouseOver={null}
                  >
                    n
                  </span>
                   === 
                  <span
                    className=" "
                    onClick={null}
                    onMouseOver={null}
                  >
                    1
                  </span>
                </span>
                 ? 
                <span
                  className=" "
                  onClick={null}
                  onMouseOver={null}
                >
                  1
                </span>
                 : 
                <span
                  className=" "
                  onClick={null}
                  onMouseOver={null}
                >
                  <span
                    className=" "
                    onClick={null}
                    onMouseOver={null}
                  >
                    n
                  </span>
                   * 
                  <span
                    className=" "
                    onClick={null}
                    onMouseOver={null}
                  >
                    <span
                      className=" "
                      onClick={null}
                      onMouseOver={null}
                    >
                      fac
                    </span>
                    (
                    <span
                      className=" "
                      onClick={null}
                      onMouseOver={null}
                    >
                      <span
                        className=" "
                        onClick={null}
                        onMouseOver={null}
                      >
                        n
                      </span>
                       - 
                      <span
                        className=" "
                        onClick={null}
                        onMouseOver={null}
                      >
                        1
                      </span>
                    </span>
                    )
                  </span>
                </span>
              </span>
              ;
            </span>
            
  }
          </span>
        </span>
      </span>
    </span>
  </code>
</pre>
`;

exports[`AnnotatedCodeView with a node highlighted 1`] = `
<pre
  className="AnnotatedCode"
  onMouseLeave={null}
>
  <code>
    <span
      className=" "
      onClick={null}
      onMouseOver={null}
    >
      <span
        className=" "
        onClick={null}
        onMouseOver={null}
      >
        <span
          className=" "
          onClick={null}
          onMouseOver={null}
        >
          function 
          <span
            className=" "
            onClick={null}
            onMouseOver={null}
          >
            fac
          </span>
          (
          <span
            className="highlighted "
            onClick={null}
            onMouseOver={null}
          >
            n
          </span>
          ) 
          <span
            className=" "
            onClick={null}
            onMouseOver={null}
          >
            {
    
            <span
              className=" "
              onClick={null}
              onMouseOver={null}
            >
              return 
              <span
                className=" "
                onClick={null}
                onMouseOver={null}
              >
                <span
                  className=" "
                  onClick={null}
                  onMouseOver={null}
                >
                  <span
                    className=" "
                    onClick={null}
                    onMouseOver={null}
                  >
                    n
                  </span>
                   === 
                  <span
                    className=" "
                    onClick={null}
                    onMouseOver={null}
                  >
                    1
                  </span>
                </span>
                 ? 
                <span
                  className=" "
                  onClick={null}
                  onMouseOver={null}
                >
                  1
                </span>
                 : 
                <span
                  className=" "
                  onClick={null}
                  onMouseOver={null}
                >
                  <span
                    className=" "
                    onClick={null}
                    onMouseOver={null}
                  >
                    n
                  </span>
                   * 
                  <span
                    className=" "
                    onClick={null}
                    onMouseOver={null}
                  >
                    <span
                      className=" "
                      onClick={null}
                      onMouseOver={null}
                    >
                      fac
                    </span>
                    (
                    <span
                      className=" "
                      onClick={null}
                      onMouseOver={null}
                    >
                      <span
                        className=" "
                        onClick={null}
                        onMouseOver={null}
                      >
                        n
                      </span>
                       - 
                      <span
                        className=" "
                        onClick={null}
                        onMouseOver={null}
                      >
                        1
                      </span>
                    </span>
                    )
                  </span>
                </span>
              </span>
              ;
            </span>
            
  }
          </span>
        </span>
      </span>
    </span>
  </code>
</pre>
`;

exports[`AnnotatedCodeView with nodes selected 1`] = `
<pre
  className="AnnotatedCode"
  onMouseLeave={null}
>
  <code>
    <span
      className=" "
      onClick={null}
      onMouseOver={null}
    >
      <span
        className=" "
        onClick={null}
        onMouseOver={null}
      >
        <span
          className=" "
          onClick={null}
          onMouseOver={null}
        >
          function 
          <span
            className=" "
            onClick={null}
            onMouseOver={null}
          >
            fac
          </span>
          (
          <span
            className=" selected"
            onClick={null}
            onMouseOver={null}
          >
            n
          </span>
          ) 
          <span
            className=" "
            onClick={null}
            onMouseOver={null}
          >
            {
    
            <span
              className=" "
              onClick={null}
              onMouseOver={null}
            >
              return 
              <span
                className=" selected"
                onClick={null}
                onMouseOver={null}
              >
                <span
                  className=" "
                  onClick={null}
                  onMouseOver={null}
                >
                  <span
                    className=" "
                    onClick={null}
                    onMouseOver={null}
                  >
                    n
                  </span>
                   === 
                  <span
                    className=" "
                    onClick={null}
                    onMouseOver={null}
                  >
                    1
                  </span>
                </span>
                 ? 
                <span
                  className=" "
                  onClick={null}
                  onMouseOver={null}
                >
                  1
                </span>
                 : 
                <span
                  className=" "
                  onClick={null}
                  onMouseOver={null}
                >
                  <span
                    className=" "
                    onClick={null}
                    onMouseOver={null}
                  >
                    n
                  </span>
                   * 
                  <span
                    className=" "
                    onClick={null}
                    onMouseOver={null}
                  >
                    <span
                      className=" "
                      onClick={null}
                      onMouseOver={null}
                    >
                      fac
                    </span>
                    (
                    <span
                      className=" "
                      onClick={null}
                      onMouseOver={null}
                    >
                      <span
                        className=" "
                        onClick={null}
                        onMouseOver={null}
                      >
                        n
                      </span>
                       - 
                      <span
                        className=" "
                        onClick={null}
                        onMouseOver={null}
                      >
                        1
                      </span>
                    </span>
                    )
                  </span>
                </span>
              </span>
              ;
            </span>
            
  }
          </span>
        </span>
      </span>
    </span>
  </code>
</pre>
`;
